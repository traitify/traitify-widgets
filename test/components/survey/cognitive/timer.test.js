/** @jest-environment jsdom */
import {act} from "react-test-renderer";
import Component from "components/survey/cognitive/timer";
import ComponentHandler from "support/component-handler";
import useContainer from "support/hooks/use-container";
import useWindowMock from "support/hooks/use-window-mock";

describe("Timer", () => {
  let component;
  let originalDate;
  let props;

  useContainer();
  useWindowMock("setTimeout");

  beforeAll(() => {
    originalDate = Date.now;
  });

  beforeEach(() => {
    const now = Date.now();
    Date.now = jest.fn().mockName("now").mockReturnValue(now);

    props = {
      onFinish: jest.fn().mockName("onFinish"),
      startTime: now,
      timeAllowed: 1500
    };
  });

  afterAll(() => {
    Date.now = originalDate;
  });

  it("renders component", async() => {
    component = await ComponentHandler.setup(Component, {props});

    expect(component.tree).toMatchSnapshot();
    expect(props.onFinish).not.toHaveBeenCalled();
  });

  it("renders component with no time remaining", async() => {
    props.startTime -= (props.timeAllowed + 10) * 1000;
    component = await ComponentHandler.setup(Component, {props});

    expect(component.tree).toMatchSnapshot();
    expect(props.onFinish).toHaveBeenCalled();
  });

  it("renders component with time passed", async() => {
    props.startTime -= 1234;
    component = await ComponentHandler.setup(Component, {props});

    expect(component.tree).toMatchSnapshot();
    expect(props.onFinish).not.toHaveBeenCalled();
  });

  it("sets time left", async() => {
    await ComponentHandler.setup(Component, {props});

    expect(Date.now).toHaveBeenCalled();
    expect(props.onFinish).not.toHaveBeenCalled();
  });

  it("submits", async() => {
    props.startTime -= props.timeAllowed * 1000;
    await ComponentHandler.setup(Component, {props});

    expect(props.onFinish).toHaveBeenCalled();
  });

  it("updates component after 1 second", async() => {
    component = await ComponentHandler.setup(Component, {props});

    expect(setTimeout).toHaveBeenCalledWith(expect.any(Function), 1000);

    Date.now.mockReturnValue(Date.now() + 1000);
    act(() => setTimeout.mock.calls[0][0]());

    expect(component.tree).toMatchSnapshot();
  });
});
