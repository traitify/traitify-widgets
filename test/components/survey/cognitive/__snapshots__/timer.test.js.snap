// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timer renders component 1`] = `
<div
  className="timer"
>
  <span
    className="traitify--icon-fa"
    dangerouslySetInnerHTML={
      {
        "__html": "<svg aria-hidden="true" focusable="false" data-prefix="fas" data-icon="clock" class="svg-inline--fa fa-clock" role="img" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><path fill="currentColor" d="M256 512C114.6 512 0 397.4 0 256S114.6 0 256 0S512 114.6 512 256s-114.6 256-256 256zM232 120V256c0 8 4 15.5 10.7 20l96 64c11 7.4 25.9 4.4 33.3-6.7s4.4-25.9-6.7-33.3L280 243.2V120c0-13.3-10.7-24-24-24s-24 10.7-24 24z"></path></svg>",
      }
    }
  />
   
  25
  :
  00
</div>
`;

exports[`Timer renders component with no time remaining 1`] = `
<div
  className="timer"
>
  <span
    className="traitify--icon-fa"
    dangerouslySetInnerHTML={
      {
        "__html": "<svg aria-hidden="true" focusable="false" data-prefix="fas" data-icon="clock" class="svg-inline--fa fa-clock" role="img" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><path fill="currentColor" d="M256 512C114.6 512 0 397.4 0 256S114.6 0 256 0S512 114.6 512 256s-114.6 256-256 256zM232 120V256c0 8 4 15.5 10.7 20l96 64c11 7.4 25.9 4.4 33.3-6.7s4.4-25.9-6.7-33.3L280 243.2V120c0-13.3-10.7-24-24-24s-24 10.7-24 24z"></path></svg>",
      }
    }
  />
   
  0
  :
  00
</div>
`;

exports[`Timer renders component with time passed 1`] = `
<div
  className="timer"
>
  <span
    className="traitify--icon-fa"
    dangerouslySetInnerHTML={
      {
        "__html": "<svg aria-hidden="true" focusable="false" data-prefix="fas" data-icon="clock" class="svg-inline--fa fa-clock" role="img" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><path fill="currentColor" d="M256 512C114.6 512 0 397.4 0 256S114.6 0 256 0S512 114.6 512 256s-114.6 256-256 256zM232 120V256c0 8 4 15.5 10.7 20l96 64c11 7.4 25.9 4.4 33.3-6.7s4.4-25.9-6.7-33.3L280 243.2V120c0-13.3-10.7-24-24-24s-24 10.7-24 24z"></path></svg>",
      }
    }
  />
   
  24
  :
  58
</div>
`;

exports[`Timer updates component after 1 second 1`] = `
<div
  className="timer"
>
  <span
    className="traitify--icon-fa"
    dangerouslySetInnerHTML={
      {
        "__html": "<svg aria-hidden="true" focusable="false" data-prefix="fas" data-icon="clock" class="svg-inline--fa fa-clock" role="img" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><path fill="currentColor" d="M256 512C114.6 512 0 397.4 0 256S114.6 0 256 0S512 114.6 512 256s-114.6 256-256 256zM232 120V256c0 8 4 15.5 10.7 20l96 64c11 7.4 25.9 4.4 33.3-6.7s4.4-25.9-6.7-33.3L280 243.2V120c0-13.3-10.7-24-24-24s-24 10.7-24 24z"></path></svg>",
      }
    }
  />
   
  24
  :
  59
</div>
`;
